app/javascript/app/document-capture/components/form-steps.jsx

```diff
@@ -0,0 +1,97 @@
+import React, { useState, useEffect } from 'react';
+import PropTypes from 'prop-types';
+
+function FormSteps({ steps, onComplete }) {
+  const [values, setValues] = useState({});
+  const [stepIndex, setStepIndex] = useState(0);
+
+  function setStepValue(name, nextStepValue) {
+    setValues({ ...values, [name]: nextStepValue });
+  }
+
+  /**
+   * Increments state to the next step, or calls onComplete callback if the
+   * current step is the last step.
+   */
+  function toNextStep() {
+    const nextStepIndex = stepIndex + 1;
+    const isComplete = nextStepIndex === steps.length;
+    if (isComplete) {
+      // Clear step parameter from URL.
+      window.history.pushState(null, null, window.location.pathname);
```

Comment:
I'd considered to pull out the logic surrounding history to something more reusable, like a `useHistoryParam` hook 
that would be responsible for managing the majority of this history syncing behavior.

Pseudo-code:

```js
const [ step, setStep ] = useHistoryParam( 'step' );
// setStep( 1 );
// setStep( null, { replace: true } );
```

It might still be worth exploring, but I encountered a few challenges along the way:

- A generic solution ideally wouldn't make assumptions about there being _at most one_ query parameter, but instead surgically operate to add, remove, or modify the single relevant query parameter. This is trivial if we could assume browser support of [URLSearchParams](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams), but we can't, as it's not supported in Internet Explorer. The alternative requires either managing the query parameters by string manipulation (error-prone) or pulling in a separate library or polyfill (extra overhead). By choosing _not_ to be general-purpose, but instead tied to this specific application of the document upload form, we can be more comfortable to impose this restriction of setting, clearing, and replacing the query string.
- In the usage here, it would likely require some mapping back and forth between the two values "step" and "name", since `useHistoryParam` would ideally only need to support one value (the query parameter key).
- It still requires some syncing logic from the component which uses the hook, where the `step` value becomes a dependency of some lifecycle hooks.